### PR TITLE
Simplify similarity module and friends

### PR DESCRIPTION
**Pull request by s1monw**
SimilarityModule was binding two different interfaces SimilaritySerivce and SimilarityLookupSerice.
Both used a class Similarities which was holding some default impls etc. Thit commit folds
all the logic into a rather simplified SimilarityService which has not construction time dependency to
any other service in the system anymore. It's soely constructued from custom similarities, the index name
and index settings and SimilarityModule is just trivial glue code with out much logic.
This also adds a simple unittest for basic test coverage of the service.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/LocalNodeId.java [hunk 1]
```diff
@@ -1,40 +0,0 @@
...
-import java.lang.annotation.Target;
-
-import static java.lang.annotation.ElementType.FIELD;
-import static java.lang.annotation.ElementType.PARAMETER;
-import static java.lang.annotation.RetentionPolicy.RUNTIME;
-
-/**
- *
- */
-@BindingAnnotation
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/LocalNodeId.java]: I love these annotations without documentation.

Comment (s1monw) [inline core/src/main/java/org/elasticsearch/index/LocalNodeId.java]: LOL yeah it's not used anywhere so I figured we trash it

**Comment by nik9000** [comment]
LGTM - didn't do a super thorough review but it looks right.

File: core/src/main/java/org/elasticsearch/index/similarity/SimilarityService.java [hunk 2]
```diff
@@ -25,55 +25,96 @@
...
+        defaults.put("default", DefaultSimilarityProvider::new);
+        defaults.put("BM25", BM25SimilarityProvider::new);
+        buildIn.put("default", DefaultSimilarityProvider::new);
+        buildIn.put("BM25", BM25SimilarityProvider::new);
+        buildIn.put("DFR", DFRSimilarityProvider::new);
+        buildIn.put("IB", IBSimilarityProvider::new);
+        buildIn.put("LMDirichlet", LMDirichletSimilarityProvider::new);
+        buildIn.put("LMJelinekMercer", LMJelinekMercerSimilarityProvider::new);
+        DEFAULTS = Collections.unmodifiableMap(defaults);
+        BUILD_IN = Collections.unmodifiableMap(buildIn);
```
Comment (rjernst) [inline core/src/main/java/org/elasticsearch/index/similarity/SimilarityService.java]: `BUILT_IN`?

**Comment by rjernst** [comment]
Fantastic! LGTM

**Pull request merged**